File: backtest-engine/internal/portfolio/portfolio.go
package portfolio

import (
	"fmt"
	"time"

	"github.com/parsedong/stock/backtest-engine/internal/matcher"
)

type lot struct {
	qty     float64
	buyDate time.Time
}

type Position struct {
	Symbol    string
	Quantity  float64
	CostPrice float64
	lots      []lot
}

func (p *Position) AvailableQty(today time.Time) float64 {
	var avail float64
	for _, l := range p.lots {
		if today.After(l.buyDate) {
			avail += l.qty
		}
	}
	return avail
}

type Portfolio struct {
	cash      float64
	positions map[string]*Position
}

func New(initCapital float64) *Portfolio {
	return &Portfolio{cash: initCapital, positions: make(map[string]*Position)}
}

func (p *Portfolio) Cash() float64 { return p.cash }

func (p *Portfolio) Position(symbol string) *Position {
	if pos, ok := p.positions[symbol]; ok {
		return pos
	}
	return &Position{Symbol: symbol}
}

func (p *Portfolio) TotalValue(prices map[string]float64) float64 {
	total := p.cash
	for sym, pos := range p.positions {
		if price, ok := prices[sym]; ok {
			total += pos.Quantity * price
		}
	}
	return total
}

func (p *Portfolio) ApplyFill(fill matcher.Fill, date time.Time) error {
	if fill.Status != matcher.Filled {
		return nil
	}
	cost := fill.FilledQty*fill.FilledPrice + fill.Commission + fill.StampDuty

	switch fill.Side {
	case matcher.Buy:
		if cost > p.cash {
			return fmt.Errorf("insufficient cash: need %.2f, have %.2f", cost, p.cash)
		}
		p.cash -= cost
		pos := p.positions[fill.Symbol]
		if pos == nil {
			pos = &Position{Symbol: fill.Symbol}
			p.positions[fill.Symbol] = pos
		}
		totalCost := pos.CostPrice*pos.Quantity + fill.FilledPrice*fill.FilledQty
		pos.Quantity += fill.FilledQty
		if pos.Quantity > 0 {
			pos.CostPrice = totalCost / pos.Quantity
		}
		pos.lots = append(pos.lots, lot{qty: fill.FilledQty, buyDate: date})

	case matcher.Sell:
		pos := p.positions[fill.Symbol]
		if pos == nil || pos.Quantity < fill.FilledQty {
			return fmt.Errorf("insufficient position for %s", fill.Symbol)
		}
		p.cash += fill.FilledQty*fill.FilledPrice - fill.Commission - fill.StampDuty
		pos.Quantity -= fill.FilledQty
		remaining := fill.FilledQty
		for i := range pos.lots {
			if remaining <= 0 {
				break
			}
			if pos.lots[i].qty <= remaining {
				remaining -= pos.lots[i].qty
				pos.lots[i].qty = 0
			} else {
				pos.lots[i].qty -= remaining
				remaining = 0
			}
		}
		if pos.Quantity == 0 {
			delete(p.positions, fill.Symbol)
		}
	}
	return nil
}

File: python/backtest/portfolio.py
from datetime import date
from backtest.types import Fill, FillStatus, OrderSide, Position, PortfolioSnapshot


class PortfolioTracker:
    """本地持仓追踪器，根据 Fill 更新现金和仓位。"""

    def __init__(self, init_cash: float):
        self._cash = init_cash
        self._positions: dict[str, _PositionDetail] = {}

    @property
    def cash(self) -> float:
        return self._cash

    def position(self, symbol: str) -> Position:
        if symbol not in self._positions:
            return Position(symbol=symbol, quantity=0.0, cost_price=0.0)
        p = self._positions[symbol]
        return Position(symbol=symbol, quantity=p.quantity, cost_price=p.cost_price)

    def apply_fill(self, fill: Fill, trade_date: date) -> None:
        if fill.status != FillStatus.FILLED:
            return

        if fill.side == OrderSide.BUY:
            cost = fill.filled_qty * fill.filled_price + fill.commission + fill.stamp_duty
            self._cash -= cost
            if fill.symbol not in self._positions:
                self._positions[fill.symbol] = _PositionDetail()
            self._positions[fill.symbol].add_lot(fill.filled_qty, fill.filled_price)

        elif fill.side == OrderSide.SELL:
            proceeds = fill.filled_qty * fill.filled_price - fill.commission - fill.stamp_duty
            self._cash += proceeds
            if fill.symbol in self._positions:
                self._positions[fill.symbol].reduce(fill.filled_qty)
                if self._positions[fill.symbol].quantity <= 0:
                    del self._positions[fill.symbol]

    def total_value(self, prices: dict[str, float]) -> float:
        total = self._cash
        for sym, pos in self._positions.items():
            total += pos.quantity * prices.get(sym, 0.0)
        return total

    def snapshot(self, prices: dict[str, float]) -> PortfolioSnapshot:
        positions = {
            sym: Position(symbol=sym, quantity=p.quantity, cost_price=p.cost_price)
            for sym, p in self._positions.items()
        }
        return PortfolioSnapshot(
            cash=self._cash,
            positions=positions,
            total_value=self.total_value(prices),
        )


class _PositionDetail:
    def __init__(self):
        self.quantity = 0.0
        self.cost_price = 0.0

    def add_lot(self, qty: float, price: float) -> None:
        total_cost = self.cost_price * self.quantity + price * qty
        self.quantity += qty
        self.cost_price = total_cost / self.quantity if self.quantity > 0 else 0.0

    def reduce(self, qty: float) -> None:
        self.quantity -= qty
